### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by dotfloat**
Resolves: #807 
Resolves: #916 
Resolves: #1011 

We've decided to go for a HTTP (non-S) server with some basic authentication system. This PR uses the standard HTTP [`Authorization: Basic [...]`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization) method of plain-text authentication.

Once the server is started, it dumps metadata about itself to disk at `[ERT project root]/storage_server.json`. This JSON file contains the server's hostname (Should make sure that this is the network-accessible hostname), port (randomly generated by Flask if you specify a port of 0) and the generated authentication token, unique for every instance. It is assumed that the storage server is running if this json file exists. (PS: There's really no way of doing any file locking on NFS or any atomic file operations, so we could end up with race conditions)

The server is now started at the beginning of ERT (when specifying `--enable-new-server`) and the `StorageClient` will patiently wait until the server is ready to accept connections before it tries to connect. To do this, I "ported" the code to use gunicorn, even when not in production.

### CODE REVIEW

**Comment by ertomatic** [comment]
Can one of the admins verify this patch?

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -29,7 +29,7 @@ jobs:
       if: matrix.os == 'ubuntu-latest'
       run: |
         sudo apt-get update
-        sudo apt-get install liblapack-dev valgrind clang cmake cmake-data xvfb libxcb-image0 libxcb-icccm4 libxcb-keysyms1 libxcb-randr0 libxcb-render0 libxcb-render-util0 libxcb-shape0 libxcb-shm0 libxcb-xfixes0 libxcb-xinerama0 libfontconfig1 libxcb-xkb1 libxkbcommon-x11-0 libdbus-1-3
+        sudo apt-get install liblapack-dev valgrind clang cmake cmake-data xvfb libxcb-image0 libxcb-icccm4 libxcb-keysyms1 libxcb-randr0 libxcb-render0 libxcb-render-util0 libxcb-shape0 libxcb-shm0 libxcb-xfixes0 libxcb-xinerama0 libfontconfig1 libxcb-xkb1 libxkbcommon-x11-0 libdbus-1-3 ninja-build
```
Comment (sondreso) [inline .github/workflows/build.yml:32]: Is ninja already installed on OSX?

Comment (dotfloat) [reply .github/workflows/build.yml:32]: I guess? The tests pass which means libres and libecl are built somehow. (Unless macOS ships with them nowadays)

**Comment by oysteoh** [comment]
If you run twice from the same runpath - you will get 
`RuntimeError: storage_server.json already exists`

i guess we should just update with a new token if it exists?

**Comment by oysteoh** [comment]
on my mac - first time i launched i was prompted with a system message telling me python wanted to open a network port on my machine - and if i wanted to allow. is this something we should expect and get our users to do? usually you need admin for this kind of stuff - not everyone has that

**Comment by oysteoh** [comment]
i'm not getting any data out from `curl -u __token__: the_token 127.0.0.1:58403/ensembles`
just `{"ensembles":[]}`

File: .github/workflows/build.yml (hunk 1 above)
Comment (sondreso) [reply .github/workflows/build.yml:32]: Interesting! I was just curious, thanks for the answer!

File: .build_install.sh [hunk 2]
```diff
@@ -1,18 +1,19 @@
...
 -DBUILD_APPLICATIONS=ON \
 -DCMAKE_INSTALL_PREFIX=$INSTALL_DIR \
 -DCMAKE_PREFIX_PATH=$INSTALL_DIR \
 -DCMAKE_INSTALL_NAME_DIR=$INSTALL_DIR/lib 
-  make install
-  popd
-  popd
+  ninja -C $src/build install
+
+  rm -rf $src
```
Comment (sondreso) [inline .build_install.sh:16]: This script has a lot of dependents in other repos, think it is best to leave out the ninja stuff :disappointed:

Comment (dotfloat) [reply .build_install.sh:16]: :cry:

**Comment by dotfloat** [comment]
> If you run twice from the same runpath - you will get
> `RuntimeError: storage_server.json already exists`
> 
> i guess we should just update with a new token if it exists?

The server should delete this file when it's closed. I'll check.

> on my mac - first time i launched i was prompted with a system message telling me python wanted to open a network port on my machine - and if i wanted to allow. is this something we should expect and get our users to do? usually you need admin for this kind of stuff - not everyone has that

I assume this happens because we're opening a server on `0.0.0.0`. We could add a check that if you're on a mac it will only open on `localhost`? Otherwise we'll need to live with the OS having a firewall. After all, we *are* starting a server.

**Comment by dotfloat** [comment]
@oysteoh test this please

**Comment by oysteoh** [comment]
when terminating `ert api`

```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/OOH/code/virtualenvs/ert/bin/ert", line 7, in <module>
    exec(compile(f.read(), __file__, 'exec'))
  File "/Users/OOH/code/github_equinor/ert/ert_shared/bin/ert", line 5, in <module>
    main()
  File "/Users/OOH/code/github_equinor/ert/ert_shared/main.py", line 338, in main
    args.func(args)
  File "/Users/OOH/code/github_equinor/ert/ert_shared/plugins/plugin_manager.py", line 295, in __exit__
    shutil.rmtree(self.tmp_dir)
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/shutil.py", line 706, in rmtree
    onerror(os.lstat, path, sys.exc_info())
  File "/usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/lib/python3.8/shutil.py", line 704, in rmtree
    orig_st = os.lstat(path)
FileNotFoundError: [Errno 2] No such file or directory: '/var/folders/n_/_9gf15_97blgklpptz48qxzw0000gp/T/tmp9tz_7ztf'
```

**Comment by oysteoh** [comment]
ert `mode` `file` --enable_new_storage gives me curl ...... 127.0.0.1: ....... to test and connect
ert api prints 0.0.0.0: .... to test and connect . 

127.0.0.1 works perfectly fine - 0.0.0.0 not so much - but of course i can change it. But i think those two should print the same

**Comment by oysteoh** [comment]
still get no results back from the api other than `{"ensembles":[]}`

**Comment by oysteoh** [comment]
running twice from same folder now works! 👍

**Comment by oysteoh** [comment]
i now get my ensembles 🍾

File: .build_install.sh (hunk 2 above)
Comment (oysteoh) [reply .build_install.sh:16]: should the ninja stuff go as @sondreso points out?

**Comment by oysteoh** [comment]
how am i - as a client of the api - supposed to retrieve the token ?

**Review by oysteoh — APPROVED** [review]

**Pull request merged**